docs/features/communication/abi_compatible_data_types/index.rst

```diff
@@ -0,0 +1,491 @@
+
+This evaluation initially targets the following process configurations:
+
+* Processes running on the same operating system.
+* Processes running on different operating systems but under the same hypervisor.
+
+Supporting different endianness between processes is explicitly out of scope, as it inherently demands bit manipulation, effectively requiring serialization.
+Different bit widths, however, are implicitly supported by specifying the width of all types and excluding word-size integers.
+
+The following data types shall be supported by the IPC mechanism:
+
+* **Primitive Types**:
+
+  * Boolean
+  * Numeric (fixed-size integers 8-128 bits, signed and unsigned; IEEE 754 floating-point numbers)
+
+* **Sequence Types**:
+
+  * Array (fixed-length)
+
+* **User-Defined Types**:
+
+  * Struct
+  * Tuple
+  * Enum (tag-only)
+  * Variant ("tagged union"; *optional*)
+
+* **Fixed-Size, Variable-Length Containers**:
+
+  * Vector
+  * Queue
+  * Hash map (*optional*)
+  * Hash set (*optional*)
+  * Binary tree (*optional*)
+
+* **Specialized Variants**:
+
+  * Result
+  * Option
+
+All provided data types must ensure fixed size and consistent memory layouts.
+
+Type Description
+----------------
+
+A critical scalability feature involves gateway processes, which subscribe to IPC endpoints and translate ABI-compatible data types into external serialization formats. These gateways require the ability to interpret data without compile-time access to type definitions. To address this, an explicit runtime-readable type description format is necessary. This description allows dynamic, runtime interpretation of data structures, enabling the addition of new IPC topics without recompiling gateway processes.
+
+In summary, the motivation behind this feature request is to define and standardize ABI-compatible data types and a runtime-accessible type description mechanism to ensure interoperability and scalability in zero-copy IPC scenarios involving multiple languages and dynamic environments.
+
+
+.. Rationale
+.. ==========
+
+
+Specification
+=============
+
+ABI Compatibility
+-----------------
+
+This specification defines the set of rules and constraints for representing data types in shared memory such that they can be interpreted consistently across processes implemented in C++17 and Rust 1.8x. These types enable zero-copy inter-process communication by enforcing ABI compatibility at the memory layout level. The focus is on data exchange between processes using the same endianness.
+
+Assumptions
+^^^^^^^^^^^
+
+* Shared memory regions are mapped at correctly aligned virtual addresses in both processes.
+* No serialization or runtime copying occurs when interpreting a type from memory.
+* Processes use the same endianness.
+* No synchronization or atomicity guarantees are defined at the data type level; these are provided by the IPC framework.
```

Comment:
As mentioned above, FFI boundary crossing is also a use case. There user code must provide the synchronization.

```suggestion
* No synchronization or atomicity guarantees are defined at the data type level; these must be provided by the accessing code.
```